### ISSUE TITLE
NOS3 Local Issue(s)

### ISSUE THREAD

**Issue by SpaceCyberCoffee** [nasa/nos3#255]
Hi,

NOS3 runs fine with the provided VMs, even with the latest Vagrant (2.4.x) and Docker (>=25.x) versions.

However I have problems with NOS3 being run locally within Ubuntu 22.04 LTS.

Even though the `make prep` and the `make` report no issues, after doing the `make launch` the terminal tabs and windows are opened, COSMOS is running, but I have the following problems:

NOS3 Flight Software terminal window shows this error:

> EVS Port1 42/1/SAMPLE 32: SAMPLE: Request device data reported error -1

In Cosmos, “Debug” and “Radio” interfaces are stuck at “Attempting” and the “NOS3 Time Driver” stops at tick = 920

And the three 42 windows are opened, but they are all black.

Any workarounds for this? 
Or I am missing something?

Thank you in advance for your support and for your amazing work with NOS3!

**Comment by jlucas9** [comment]
Hi @SpaceCyberCoffee ,

Thanks for submitting the issue. I can confirm a fresh Ubuntu 22.04 results in similar problems for me:
![image](https://github.com/nasa/nos3/assets/6588622/4cf282a3-c1e3-4ac2-b480-e6503a878a1f)

Appears the bug was in the sample simulator. It was crashing then leaving the entire system to hang. Can you give the `nos3#255-Ubuntu2204` branch a try and see if it fixes it for you?

**Comment by SpaceCyberCoffee** [comment]
Hi @jlucas9 thank you for your Ubuntu2204 branch!
I have tested it and it solved most of the issues I had before!
I am only left with two problems: Debug and Radio interfaces in COSMOS are still stuck at "Attempting", but both 42 and the Flight Software are running!

The error messages for "Radio" and "Debug" are the following (from COSMOS Log Messages):
```
RADIO: Connection Failed: SocketError : Hostname not known: radio_sim
DEBUG: Connection Failed: SocketError : Hostname not known: nos_fsw
```

Should I manually modify the Hostnames somewhere?

Thank you again for the dedicated branch!

**Comment by jlucas9** [comment]
@SpaceCyberCoffee - I'm happy to help!
Did you install cosmos using the `make prep` or manually? 

I'd recommend doing a `make stop-gsw` followed by `rm -r ~/.nos3` and then `make prep` again just to be sure the version of OpenC3 cosmos with ports already open for you is running.

Here are the changes are from the main OpenC3 repository if you wanted to manually recreate things above:
* https://github.com/nasa-itc/openc3-nos3/commit/1b503cd57eff5361bb1182952b875a56079d48a8

**Comment by SpaceCyberCoffee** [comment]
Hi again @jlucas9, thank you for your quick reply!
I initially used `make prep`, then I followed your instructions (`make stop-gsw` -> `rm -r ~/.nos3` -> `make prep`), but I still have the same issue as before.

I also did a `make clean` and did all the steps again.

However, upon the completion of the `make prep` command, the interfaces are already popping-up in Cosmos, with Radio and Debug being stuck in "Attempting".

For what it is worth, COSMOS did not prompt for a new password.

Indeed, I have noticed the following message while after doing the `make prep`:

> WARN[0000] a network with name openc3-cosmos-network exists but was not created for project "cosmos".
> Set `external: true` to use an existing network

**Comment by jlucas9** [comment]
Hi @SpaceCyberCoffee,

I double checked following those same steps myself and things seem happy. Do you see any other errors reported during the `make` process to build everything?

![image](https://github.com/nasa/nos3/assets/6588622/55999ce6-cec8-4f27-95c1-3e65cddc3ce8)

One thought I had was maybe the groups your local user are a part of could be preventing editing the docker boxes that COSMOS runs in. Here is what I am in now:
```
nos3@nos3-VirtualBox:~/git/n0$ groups
nos3 adm cdrom sudo dip plugdev lpadmin lxd sambashare docker vboxsf
```

Note that I am running a windows host with Ubuntu 22 in a VM and my nos3 folder shared in - this required the additional `vboxsf` group.

**Comment by SpaceCyberCoffee** [comment]
Hey @jlucas9 

This is the output of the `make launch`. 
No errors in the terminal are reported.

![nos3_makelaunch](https://github.com/nasa/nos3/assets/159529204/0c326692-57ae-4e5d-b38d-cba3110c77da)

On the other hand, there is error in the NOS3 Engine Server terminal tab:
![nos3_engine](https://github.com/nasa/nos3/assets/159529204/4901ce15-2b80-4009-8346-2f0b64b1df93)

**Comment by SpaceCyberCoffee** [comment]
Hey @jlucas9,
the `make launch` command doesn't return any error.
![nos3_makelaunch](https://github.com/nasa/nos3/assets/159529204/30056d69-2206-4bd7-8a08-2cb847a497cc)

However there is the following error in the NOS3 Engine Server terminal tab:

![nos3_engine](https://github.com/nasa/nos3/assets/159529204/4efebde2-2c8a-4371-9b3b-f6186a5574da)

**Comment by SpaceCyberCoffee** [comment]
Hi @jlucas9, just a short update.

I deleted all the NOS3 and OpenC3 COSMOS Docker volumes, containers and images.
I deleted the ~/.nos3 folder and re-downloaded the feature branch #255 from scratch.

This time, at the end of the make prep, "Debug" and "Radio" are not within the Cosmos interfaces, that is empty (see screenshot).
![Uploading Screenshot 2024-02-21 171440.png…]()


But unfortunately after running the make command, that gives no errors, I am back to "Debug" and "Radio" being stuck at "Attempting", with the usual error message(s).

**Comment by jlucas9** [comment]
Hi @SpaceCyberCoffee,
I get similar errors in NOS Engine but not identical to what you reported:

![image](https://github.com/nasa/nos3/assets/6588622/6df3bd73-b7ea-4bcd-8fa4-6eb29610a239)

Does time appear to tic forward for you after you `make launch`?

You should be able to see this in the NOS3 Time Driver tab and in the 42 windows:
![image](https://github.com/nasa/nos3/assets/6588622/7addb8ef-854a-4d47-b3dc-9ae3b470cd57)

Assuming that's happy - I'm thinking maybe the network connections between the docker containers is mad. While everything is still running after `make launch`, can you `docker network ls`?

![image](https://github.com/nasa/nos3/assets/6588622/d0700b1a-fcda-4479-a9bf-ceddc4963eb9)

Additionally, I receive the following for a `docker network inspect nos3_sc_1`

![image](https://github.com/nasa/nos3/assets/6588622/8284873d-e436-4719-9508-d85bff3acc06)

**Comment by jlucas9** [comment]
Hi @SpaceCyberCoffee,
This seems to be the same thing captured in #242. The quick fix to get you fully functional is to edit the [./scripts/docker_launch.sh Line 84](https://github.com/nasa/nos3/blob/46e9ed27357891342179430682dbc917ef967b60/scripts/docker_launch.sh#L84) to use the `-` instead of the `_` when it's doing the network assignments. 
Hopefully this does it for you!

**Comment by SpaceCyberCoffee** [comment]
Hi @jlucas9,
indeed it did the trick!
I had tried before but I hadn't realize that there were 2 underscores `cosmos**_**openc3-operator**_**1`, and not only the one before the "1"!

I am happy to report that now NOS3 works flawlessly on Ubuntu 22.04.4 LTS, with Docker version 25.0.3.

Unrelated note: I have also been playing around with the latest version of `42`.
I saw that you have are planning to implement it (#207 ) in the future release.

I am happy to test it if you want and/or to share my attempt!

**Comment by jlucas9** [comment]
Just merged in #207 to the latest `dev` branch @SpaceCyberCoffee if you wanted to give it a shot!

You'll need to `make real-clean prep all` to:
* real-clean
  * Clean up all the old stuff hanging around
* prep
  * Download the latest and greatest
* all
  * Build everything

Please feel free to file more issues as you find them!

**Comment by SpaceCyberCoffee** [comment]
Hi @jlucas9,
I pulled the new `dev`, then edited the `scripts/docker_launch.sh` file to have `cosmos-openc3-operator-1`, and I followed your instructions, but now I get these errors, at the end of "Install the project.." part:

```
make[1]: Leaving directory '/home/user-dir/nos3_local/nos3'
make gsw
make[1]: Entering directory '/home/user-dir/nos3_local/nos3'
./cfg/build/gsw_build.sh
make[1]: ./cfg/build/gsw_build.sh: Permission denied
make[1]: *** [Makefile:88: gsw] Error 127
make[1]: Leaving directory '/home/user-dir/nos3_local/nos3'
make: *** [Makefile:46: all] Error 2
```
